### PR TITLE
Add unit tests to support #838

### PR DESCRIPTION
**Pull request by k-wall**
### Type of change

_Select the type of your PR_

- Enhancement

### Description

Unit tests to accompany fix for #838.  

Verified that the new test fails in the expected way if you backout the fix for #838.

Screenshot 2023-12-22 at 12 57 04

### Additional Context

_Why are you making this pull request?_

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [x] Write tests
- [ ] Make sure all tests pass
- [ ] Review performance test results. Ensure that any degradations to performance numbers are understood and justified.
- [ ] Make sure all Sonar-Lint warnings are addressed or are justifiably ignored.
- [ ] Update documentation
- [ ] Reference relevant issue(s) and close them after merging
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

### CODE REVIEW

File: kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/inband/InBandKeyManagerTest.java [hunk 1]
```diff
@@ -587,6 +592,61 @@ void shouldEncryptRecordHeadersForMultipleRecords() throws ExecutionException, I
...
+    void decryptPreservesRecordOrder() {
+        var topic = "topic";
+        var partition = 1;
+
+        var kmsService = UnitTestingKmsService.newInstance();
+        InMemoryKms kms = kmsService.buildKms(new UnitTestingKmsService.Config());
+        var kekId1 = kms.generateKey();
+        var kekId2 = kms.generateKey();
+
+        var spyKms = Mockito.spy(kms);
```
Comment (k-wall) [inline kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/inband/InBandKeyManagerTest.java:605]: 😱

Comment (k-wall) [reply kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/inband/InBandKeyManagerTest.java:605]: Hideous .. I'm hoping once we've got #774 etc we'd be able to rewrite this tests using mocks, and avoid the scary spying nonsense.

File: kroxylicious-kms-provider-kroxylicious-inmemory/src/main/java/io/kroxylicious/kms/provider/kroxylicious/inmemory/InMemoryEdek.java [hunk 2]
```diff
@@ -10,13 +10,13 @@
 import java.util.Objects;
 import java.util.UUID;
 
-record InMemoryEdek(
+public record InMemoryEdek(
```
Comment (k-wall) [inline kroxylicious-kms-provider-kroxylicious-inmemory/src/main/java/io/kroxylicious/kms/provider/kroxylicious/inmemory/InMemoryEdek.java:13]: We will be able to correct the revert the visibility change too.

File: kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/inband/InBandKeyManagerTest.java [hunk 3]
```diff
@@ -76,23 +79,30 @@ void shouldEncryptRecordValue() {
...
-        record.value().rewind();
-        assertEquals(1, encrypted.size());
-        assertNotEquals(initial, encrypted);
-        // TODO add assertion on headers
+        assertThat(encrypted.iterator())
+                .toIterable()
+                .singleElement()
+                .extracting(TestingRecord::value)
+                .extracting(ByteBuffer::array)
+                .isNotEqualTo(value);
```
Comment (SamBarker) [inline kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/inband/InBandKeyManagerTest.java:94]: I could just be me but this is a LOT clearer to read.

File: kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/inband/InBandKeyManagerTest.java (hunk 1 above)
Comment (SamBarker) [reply kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/inband/InBandKeyManagerTest.java:664]: Ohh that might be the forcing function I need to to actually merge #774

**Review by SamBarker — APPROVED** [review]
LGTM,

Approved assuming the build is fixed by cosmetic changes.

**Comment by k-wall** [comment]
Merged despite the coverage gate failure.  The coverage gate failed because of the need to increase the visibility of the InMemory* record classes.  Hopefully, we'll be able to revert this part of the change in future.

**Pull request merged**